OpenRA.Mods.Common/Traits/World/WeatherOverlay.cs

```diff
@@ -0,0 +1,391 @@
+
+		[Desc("The value that particles swing to the side each update. SwingSpeed min. and max. value.")]
+		public readonly int[] SwingSpeed = { 1, 25 };
+
+		[Desc("The value range that can be swung to the left or right. SwingAmplitude min. and max. value.")]
+		public readonly int[] SwingAmplitude = { 10, 15 };
+
+		[Desc("The colors for the particles.",
+			"For snow and rain use the temperate.pal with an index of 128-131 and 164-167.",
+			"and for a sandstorm at Dune2k the d2k.pal with an index of 78, 82, 85, 89.")]
+		public readonly Color[] ParticleColors = {
+			Color.FromArgb(255, 236, 236, 236),
+			Color.FromArgb(255, 228, 228, 228),
+			Color.FromArgb(255, 208, 208, 208),
+			Color.FromArgb(255, 188, 188, 188)
+		};
+
+		[Desc("Works only with line enabled and will fade out the tail of the line.")]
+		public readonly byte LineTailAlphaValue = 200;
+
+		public object Create(ActorInitializer init) { return new WeatherOverlay(init.World, this); }
+	}
+
+	public class WeatherOverlay : ITick, IPostRender
+	{
+		readonly WeatherOverlayInfo info;
+		readonly World world;
+		public struct Particle
+		{
+			public float PosX;
+			public float PosY;
+			public int Size;
+			public float DirectionScatterX;
+			public float Gravity;
+			public float SwingOffset;
+			public float SwingSpeed;
+			public bool SwingDirection;
+			public float SwingAmplitude;
+			public Color Color;
+			public Color TailColor;
+
+			public Particle(
+				float posX,
+				float posY,
+				int size,
+				float directionScatterX,
+				float gravity,
+				float swingOffset,
+				float swingSpeed,
+				bool swingDirection,
+				float swingAmplitude,
+				Color color,
+				Color tailColor)
+			{
+				PosX = posX;
+				PosY = posY;
+				Size = size;
+				DirectionScatterX = directionScatterX;
+				Gravity = gravity;
+				SwingOffset = swingOffset;
+				SwingSpeed = swingSpeed;
+				SwingDirection = swingDirection;
+				SwingAmplitude = swingAmplitude;
+				Color = color;
+				TailColor = tailColor;
+			}
+		}
+
+		public List<Particle> ParticleList = new List<Particle>();
+		int maxParticleCount;
+
+		enum ParticleCountFaderType { Hold, FadeIn, FadeOut }
+		ParticleCountFaderType particleCountFader = ParticleCountFaderType.FadeIn;
+
+		float targetWindXOffset = 0f;
+		float currentWindXOffset = 0f;
+		int currentWindIndex = 0;
+		long windTickCountdown = 1500;
+		float2 antiScrollPrevTopLeft;
+
+		public WeatherOverlay(World world, WeatherOverlayInfo info)
+		{
+			this.info = info;
+			this.world = world;
+			currentWindIndex = info.WindStrength.Length / 2;
+			targetWindXOffset = info.WindStrength[0];
+			maxParticleCount = CalculateParticleCount(Game.Renderer.Resolution.Width, Game.Renderer.Resolution.Height);
+		}
+
+		int CalculateParticleCount(int x, int y)
+		{
+			var tempfactor = 0;
+			if (info.ParticleDensityFactor != 0)
+				tempfactor = info.ParticleDensityFactor;
+			else tempfactor = 1;
+
+			return (x * y) / tempfactor;
```

Comment:
You could actually simplify the method:
`return (x * y) / (info.ParticleDensityFactor != 0 ? info.ParticleDensityFactor : 1)`